tools/tslint-rules/lightweightTokensRule.ts

```diff
@@ -0,0 +1,198 @@
+ */
+export class Rule extends Lint.Rules.TypedRule {
+  applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+    return this.applyWithFunction(sourceFile, checkSourceFileForLightweightTokens,
+        this.getOptions().ruleArguments as RuleArguments, program.getTypeChecker());
+  }
+}
+
+
+/**
+ * Checks whether the given source file has classes with DI constructor parameters
+ * for which lightweight tokens are suitable (for optimized tree shaking).
+ */
+function checkSourceFileForLightweightTokens(ctx: Context, typeChecker: ts.TypeChecker): void {
+  // Relative path for the current TypeScript source file. This allows for
+  // relative globs being used in the rule options.
+  const relativeFilePath = path.relative(process.cwd(), ctx.sourceFile.fileName);
+  const [enabledFilesGlobs] = ctx.options;
+  const visitNode = (node: ts.Node) => {
+    if (ts.isClassDeclaration(node)) {
+      checkClassDeclarationForLightweightToken(node, ctx, typeChecker);
+    } else {
+      ts.forEachChild(node, visitNode);
+    }
+  };
+
+  if (enabledFilesGlobs.some(g => minimatch(relativeFilePath, g))) {
+    ts.forEachChild(ctx.sourceFile, visitNode);
+  }
+}
+
+/**
+ * Checks whether the given class declarations has a constructor with parameters
+ * for which lightweight tokens are suitable (for optimized tree shaking).
+ */
+function checkClassDeclarationForLightweightToken(
+    node: ts.ClassDeclaration, ctx: Context, typeChecker: ts.TypeChecker) {
+  // If a class has any decorator, it is assumed to be an Angular decorator.
+  // There are no other decorators available and we do not want to complicate
+  // this lint rule.
+  if (!node.decorators || node.decorators.length === 0) {
+    return;
+  }
+
+  for (const member of node.members) {
+    if (member.kind === ts.SyntaxKind.Constructor) {
+      checkConstructorForLightweightTokens(member as ts.ConstructorDeclaration, ctx, typeChecker);
+    }
+  }
+}
+
+/**
+ * Checks whether the given constructor has parameters where a lightweight
+ * token is suitable.
+ */
+function checkConstructorForLightweightTokens(
+    node: ts.ConstructorDeclaration, ctx: Context, typeChecker: ts.TypeChecker) {
+  for (const param of node.parameters) {
+    // Skip parameters without an explicit type. This should never happen in this
+    // repository but we handle it silently if there are such instances.
+    if (param.type === undefined) {
+      continue;
+    }
+
+    // Skip parameters which already have `@Inject` or are not using `@Optional` at all.
+    const {isOptional, hasInject} = analyzeDecorators(param);
+    if (hasInject || !isOptional) {
+      continue;
+    }
+
+    // Determine the type being used for injection of this parameter.
+    const injectionType = getInjectionType(param.type);
+    if (injectionType === null) {
+      continue;
+    }
+
+    let symbol = typeChecker.getSymbolAtLocation(injectionType);
```

Comment:
Can we avoid going through the type checker here? From what I can tell, it's only used to resolve aliases for the ignored list, but we don't tend to use aliases that much in this repository. The advantage is that it should be slightly quicker and it can run inside of an IDE.

Reply:
I think we need it for more. i.e. checking whether the referenced class is declared in a type definition, or if the referenced class resolves to the containing class (that could be feasible without the checker).

I agree that type checker has the downside of being little bit more slow, but I don't think there is a good way around it. The type checker instance should be set up anyway due to other rules (with internally cached symbol links). IDE's should be fine with type checker rules as long as your IDE uses the correct tsconfig (which WebStorm does IIRC).

Reply:
My concern was mostly for IDEs since vscode doesn't run lint rules that use the type checker.

Reply:
I think VSCode actually supports type checker/program rules. https://github.com/Microsoft/vscode-typescript-tslint-plugin#differences-with-the-vscode-tslint-extension. I'm not too deep into VSCode though, but this looks like an official Microsoft implementation using the language service.

If an IDE does not support type checker, then it's probably not too bad either. The rule will just not run, but CI will fail. I don't think there is a good way around it. We have a couple of other rules using the checker too